### ISSUE TITLE
create-diff-object fails with SIGSEGV

### ISSUE THREAD

**Issue by manojh3012** [dynup/kpatch#1193]
- Failure because of accessing struct member `sec->secsym->include` when `sec->secsym` is NULL.
- This error is new as we recetly switched our distro to use binutils 2.36 from binutils 2.31
- With binutils 2.31, `sec->secsym` seems always allocated(__Working case__)
- With binutils 2.36, `sec->secsym` allocation doesn't always seem to happen(__Failing case__)
- Further exploring the output of `readelf -Ws` on the patched objs, I found that the working case has more number of symbols than the failing case. Some optimization to reduce the number of `Section Symbols` happened between 2.31 and 2.36, is my best guess at this point(attached the full output of the readelf command in 2 cases as well)
- Failing case: 
```
readelf -s patched/fs/namespace.o | wc
    497    3813   35450
```
- Working case: 
```
readelf -s patched/fs/namespace.o | wc
    634    4772   43670
```
[readelf_symtab_2_36.txt](https://github.com/dynup/kpatch/files/6658423/readelf_symtab_2_36.txt)
[readelf_symtab_2_31.txt](https://github.com/dynup/kpatch/files/6658405/readelf_symtab_2_31.txt)

Please let me know if you need any other debug data here.

And, from a solution perspective, 
- I suppose I could go ahead and prefix every access to `sec->secsym->include` with a check for `sec->secsym`. While this would fix the SIGSEGV, I don't know enough to say that this would not cause other issues.
- I preferably would not like to move back to binutils 2.31 just for this.
- Any other ideas?

**Comment by sm00th** [comment]
Hi, we seem to have a proposed fix for the same issue in #1186. Could you try that fix? 

Also, if possible, could you please upload orig and patched .o files that trigger this issue and share a sample patch that results in those?

**Comment by manojh3012** [comment]
The failure in my case is in [kpatch_regenerate_special_section](https://github.com/dynup/kpatch/blob/v0.9.1/kpatch-build/create-diff-object.c#L2139). The patch in #1186 fixes [kpatch_include_debug_sections](https://github.com/dynup/kpatch/blob/v0.9.1/kpatch-build/create-diff-object.c#L2273). There are several places in the kpatch codebase that writes to `sec->secsym->include`. Some do with and some without a check

Sample patch below and also uploaded the orig and patched.o files with extension `.txt`. Github doesn't allow me to upload a `.o` file

```
diff --git a/fs/namespace.c b/fs/namespace.c
index 741f40c..6a38668 100644
--- a/fs/namespace.c
+++ b/fs/namespace.c
@@ -1475,6 +1475,14 @@ static int do_umount(struct mount *mnt, int flags)
 	if (retval)
 		return retval;
 
+        printk("kpatch process: %s unmounting %s from %s\n", current->comm, mnt->mnt_devname, mnt->mnt_mountpoint->d_iname);
+        if (current->real_parent) {
+            printk("real parent: %s\n", current->real_parent->comm);
+        }
+        if (current->parent) {
+            printk("parent: %s\n", current->parent->comm);
+        }
+
 	/*
 	 * Allow userspace to request a mountpoint be expired rather than
 	 * unmounting unconditionally. Unmount only happens if:
@@ -2723,6 +2731,7 @@ long do_mount(const char *dev_name, const char __user *dir_name,
 	struct path path;
 	unsigned int mnt_flags = 0, sb_flags;
 	int retval = 0;
+        char kname[PATH_MAX];
 
 	/* Discard magic */
 	if ((flags & MS_MGC_MSK) == MS_MGC_VAL)
@@ -2740,6 +2749,11 @@ long do_mount(const char *dev_name, const char __user *dir_name,
 	if (retval)
 		return retval;
 
+        if ( strncpy_from_user(kname, dir_name, PATH_MAX) > 0 ) {
+            printk("kpatch process: %s mounting %s to %s\n", current->comm, dev_name, kname);
+        }
+
+
 	retval = security_sb_mount(dev_name, &path,
 				   type_page, flags, data_page);
 	if (!retval && !may_mount())

```
[patched.namespace.o.txt](https://github.com/dynup/kpatch/files/6665115/patched.namespace.o.txt)
[orig.namespace.o.txt](https://github.com/dynup/kpatch/files/6665116/orig.namespace.o.txt)

**Comment by manojh3012** [comment]
kpatch was compiled with gcc 5.2(for various legacy reasons) while binutils moved from 2.31 to 2.36.

I narrowed down the [binutils commit](https://sourceware.org/git/gitweb.cgi?p=binutils-gdb.git;h=d1bcae833b32f1408485ce69f844dcd7ded093a8) that removes unused section symbols. In my case, I patched binutils 2.36 with the following diff which seems to workaround the problem. I need to do more testing though.
```
+Index: git/bfd/elfxx-x86.h
+===================================================================
+--- git.orig/bfd/elfxx-x86.h
++++ git/bfd/elfxx-x86.h
+@@ -19,7 +19,7 @@
+    MA 02110-1301, USA.  */
+ 
+ /* Don't generate unused section symbols.  */
+-#define TARGET_KEEP_UNUSED_SECTION_SYMBOLS FALSE
++#define TARGET_KEEP_UNUSED_SECTION_SYMBOLS TRUE
+ 
+ #include "sysdep.h"
+ #include "bfd.h"
```

**Comment by jpoimboe** [comment]
We noticed upstream that binutils 2.36 has been aggressive about removing section symbols where it can.  It caused some issues with objtool as well.

**Comment by sm00th** [comment]
> I narrowed down the [binutils commit](https://sourceware.org/git/gitweb.cgi?p=binutils-gdb.git;h=d1bcae833b32f1408485ce69f844dcd7ded093a8) that removes unused section symbols. In my case, I patched binutils 2.36 with the following diff which seems to workaround the problem. I need to do more testing though.

Thanks for pointing this out. Looks like adding secsym checks is the correct approach here. I'll make sure it doesn't break anything and will post a fix afterwards.
